runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -461,38 +517,47 @@ end
 
 --- Helper function to iterate through all of the diagnostic lines
 ---@return table list of diagnostics
-local _iter_diagnostic_lines = function(start, finish, step, bufnr, opts, client_id)
-  if bufnr == nil then
-    bufnr = vim.api.nvim_get_current_buf()
-  end
-
+local _iter_diagnostic_lines = function(cursor_position, search_forward, bufnr, opts, client_id)
```

Comment:
I don't think it makes sense anymore for this function to be iter_diagnostic_lines, I think we should probably rename it something like iter_diagnostic_position, but it might make sense to have an iter_diagnostic_lines function as well (instead of replacing the old one).

Reply:
Maybe the new function should be iter_diagnostic_range, and instead of having a "search_forward" param have a start/end param that correspond to the buf position.

Reply:
> I don't think it makes sense anymore for this function to be iter_diagnostic_lines, I think we should probably rename it something like iter_diagnostic_position

i realise this can return a single diagnostic instead of a list of diagnostics, so i renamed it to `next_diagnostic` instead. What do you think?

> it might make sense to have an iter_diagnostic_lines function as well (instead of replacing the old one)

this function is internal and doesn't look like it's used anywhere else, so i'll prefer if we can avoid keeping the old one

> Maybe the new function should be iter_diagnostic_range, and instead of having a "search_forward" param have a start/end param that correspond to the buf position.

not sure if i understand this. The previous implementation supported wrapping by scanning through lines twice (eg. curr_line+1 to last_line, 1 to curr_line-1). I updated the implementation to use a offset+modulo to iterate through positions more easily, `search_forward` here is just a boolean switch to handle the various cases. For example, if we are searching forward and a diagnostic appears on the right of the same line, the nearest diagnostic to the right is returned